### ISSUE TITLE
Insert Link Dialog from HTML editor does not focus on URL field upon open

### ISSUE THREAD

**Issue by alisterscott** [Automattic/wp-calypso#25860]
#### Steps to reproduce
1. Starting at URL: https://wordpress.com/post
2. Switch to HTML mode
3. Highlight text and choose the "link" editor item

#### What I expected

Dialog to focus on URL field (as per visual editor)

#### What happened instead

No focus

#### Browser / OS version

Chrome on macOS

#### Screenshot / Video

![insert link focus](https://user-images.githubusercontent.com/128826/42258503-6b312954-7f9f-11e8-955f-4a82d5ae63cd.gif)

#### Context / Source
 #dogfooding

**Comment by alisterscott** [comment]
I tried fixing this by adding `autoFocus={ true }` like so:

<FormFieldset>
					<FormLabel htmlFor="link_url">{ translate( 'URL' ) }</FormLabel>
					<FormTextInput
						name="link_url"
						autoFocus={ true }
						onChange={ this.setLinkUrl }
						ref={ this.bindLinkUrlRef }
						value={ linkUrl }
					/>
</FormFieldset>

but this gives a `jsx-a11y/no-autofocus` eslint violation.

Any ideas?

**Comment by ramonjd** [comment]
I've experimented quickly with some alternatives, rifled through the code, and 🤔accordingly. 

The attribute `autoFocus` appears frequently, and when it does, we've often included the appropriate comment to ignore it:

`autoFocus={ false } // eslint-disable-line jsx-a11y/no-autofocus`

The inevitable question is 'is this okay?' In this specific case I'd be bold and argue 'yes' since:

- The modal takes over as main content area, demanding the user's attention
- There's a fairly obvious sequence of user-controlled events bridging the click to the input  1) I want to add a link 2) That link is... 

I know it doesn't mean it's okay everywhere.

#24504 lists a tome of eslint errors we need to address, and our use of un-ignored `autoFocus` is one of them. Have we discussed our attitude to `autoFocus` anywhere before? I did some poking around and couldn't find anything explicit.

@ryelle Any ideas or guidance here?

As for the alternative, focussing using javascript and refs, it's the parent component, [EditorHtmlToolbar](https://github.com/Automattic/wp-calypso/blob/0296df3/client/post-editor/editor-html-toolbar/index.jsx
) that determines the visibility of the AddLinkDialog modal, and therefore the input field we'd like to auto focus. Not much happens in between. 

If I were compelled to choose, I'd opt for extending the [form text input](https://github.com/Automattic/wp-calypso/blob/0296df3/client/components/forms/form-text-input/index.jsx) component to accept an `shouldAutoFocus` prop and leave it to focus itself onMount. We might use the opportunity to update to `this.createRef()` where we can too!

Whatever the consensus, happy to contribute to a solution if I've inadvertently added bloat to this whole issue!  :)
